### ISSUE TITLE
Block future requests after `Capybara::Server#wait_for_pending_requests`

### ISSUE THREAD

**Issue by jonleighton** [jnicklas/capybara#1716]
I just had a play with [the request counter](https://github.com/jnicklas/capybara/commit/998350091be30fd84e102755f009304cc45e26dc) implemented in Capybara 2.7. It was interesting to me because we have had something quite similar in a test suite I work with for a while, so I thought it could be an opportunity to remove that custom code.

Our implementation looks like this (it's not CPU-threadsafe but we don't care because we're on MRI):

``` ruby
class RequestCounter
  @count   = 0
  @blocked = false

  class << self
    attr_accessor :count
    attr_accessor :blocked
  end

  def self.block
    self.blocked = true
    yield
  ensure
    self.blocked = false
  end

  def initialize(app)
    @app = app
  end

  def call(env)
    if self.class.blocked
      [500, {}, ["RequestCounter is blocked."]]
    else
      begin
        self.class.count += 1
        @app.call(env)
      ensure
        self.class.count -= 1
      end
    end
  end
end

Capybara.app = RequestCounter.new(Capybara.app)
```

We also use [database_cleaner](https://github.com/DatabaseCleaner/database_cleaner), and a long time ago I made sure that it would hook into `RequestCounter` to block requests while running `DatabaseCleaner.clean`:

``` ruby
RSpec.configure do |config|
  config.after(:each) do
    unless self.class.use_transactional_fixtures
      RequestCounter.block do
        sleep 0.1 until RequestCounter.count == 0
        DatabaseCleaner.clean
      end
    end
  end
end
```

I ripped out all of this to try the Capybara implementation, but unfortunately I saw a database deadlock while running my test suite.

I think what happens is something like this (note I'm using the [Poltergeist](https://github.com/teampoltergeist/poltergeist) driver):

* In the `after` hook, `wait_for_pending_requests` is called, but there are *currently* no pending requests
* Some time shortly *after* this, my page submits an XHR request to the app. Note that Poltergeist does [release](https://github.com/teampoltergeist/poltergeist/blob/master/lib/capybara/poltergeist/client/browser.coffee#L22) the page as part of its `reset` implementation, but I suspect there is some asynchronicity here.
* In the main Ruby thread, we start running `DatabaseCleaner.clean`
* Meanwhile, in the server thread, the app starts processing the XHR request, which involves some database work
* Because of the nature of the database queries running in the two threads, the database throws a deadlock error (`PG::TRDeadlockDetected: ERROR:  deadlock detected`) and my test fails

(FWIW, yes my `DatabaseCleaner.clean` code is definitely running after Capybara's [after hook](https://github.com/jnicklas/capybara/blob/master/lib/capybara/rspec.rb#L19-L24).)

To prevent this, I think `Capybara::Server` could be modified to ignore requests which come before [the before hook](https://github.com/jnicklas/capybara/blob/master/lib/capybara/rspec.rb#L25-L31) or after [the after hook](https://github.com/jnicklas/capybara/blob/master/lib/capybara/rspec.rb#L19-L24). Thus ensuring that requests will only ever hit the Rails app while we're actually in the process of running a test.

**Comment by jonleighton** [comment]
> (FWIW, yes my DatabaseCleaner.clean code is definitely running after Capybara's after hook.)

Sorry! I'm wrong about this :( Turns out my understanding of Capybara's ordering was wrong, as this test demonstrates:

``` ruby
require "rspec"

RSpec.describe "foo" do
  after { puts "after 1" }
  after { puts "after 2" }

  example "foo" do
  end
end
```

Outputs:

```
after 2
after 1
.

Finished in 0.00053 seconds (files took 0.08948 seconds to load)
1 example, 0 failures
```

I'm going to close this and try again.

**Comment by twalpole** [comment]
@jonleighton Great that someone else is looking into this, since we've had some people who think it's causing issues (Issue #1715) but I don't see how it could be (maybe they have order issues with the database_cleaner fater block too).  An important thing to notice is that wait_for_pending_requests is only called after the session has set the page back to about:blank so there really shouldn't be any more requests started after that.

**Comment by jonleighton** [comment]
Ok, here's what I have found out:

1. As demonstrated above, RSpec after hooks run in the *reverse* of the order that they are defined. The last one runs first.
2. I get deadlocks and other race conditions if my `DatabaseCleaner.clean` hook runs before Capybara's reset hook.
3. Therefore, I need to define my `DatabaseCleaner.clean` hook before `capybara/rspec` is defined
4. This is actually more difficult to achieve than I first realised, because [rspec-rails actually requires `capybara/rspec`](https://github.com/rspec/rspec-rails/blob/master/lib/rspec/rails/vendor/capybara.rb), therefore I also need to make sure my `DatabaseCleaner.clean` hook is defined before I even require `rspec/rails` (so I need to require just `rspec`, then define the `DatabaseCleaner.clean` hook, then require `rspec/rails`)

So it seems to be that it is *possible* to have a configuration which doesn't cause race conditions. However it is certainly not easy, and I wouldn't be surprised if lots of users get tripped up by these subtle issues.

(I just found RSpec's [`append_after` method](http://rspec.info/documentation/3.4/rspec-core/RSpec/Core/Configuration.html#append_after-instance_method) which looks like it'll allow me to define my `DatabaseCleaner.clean` hook after `capybara/rspec` is required, so this is an alternative way.)

I'm not sure what Capybara itself can do to mitigate this problem. Fundamentally, we want some way to ensure that Capybara's `after` hook is the first one to run, so that it definitely runs before `DatabaseCleaner` (if the user is using that). But I'm not sure there's actually a way to achieve that.

I guess documentation could help.
